espnet/nets/pytorch_backend/k2_ctc.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+
+# Copyright (c)  2021  Mobvoi Inc. (authors: Yaguang Hu)
+# Apache 2.0
+
+from functools import lru_cache
+from typing import List
+
+import torch
+import k2
+import math
+from typing import Tuple
+import logging
+
+
+class K2CTCLoss(torch.nn.Module):
+    def __init__(
+        self,
+        odim: int,
+        reduction: str = "sum",
+        device: torch.device = torch.device("cpu"),
+    ) -> None:
+        torch.nn.Module.__init__(self)
+        self.device = device
+        self.reduction = reduction
+        if reduction not in ["sum", "none"]:
+            logging.error(
+                f"k2 ctc loss reduction type: {reduction} is not supported yet,"
+                "change it to be 'sum'"
+            )
+            self.reduction = "sum"
+        self.graph_compiler = CtcTrainingGraphCompiler(odim, device=device)
+
+    def forward(
+        self,
+        log_probs: torch.Tensor,
+        targets: torch.Tensor,
+        input_lengths: torch.Tensor,
+        target_lengths: torch.Tensor,
+    ) -> torch.Tensor:
+        log_probs = log_probs.permute(
+            1, 0, 2
+        )  # now log_probs is [N, T, C]  batchSize x seqLength x alphabet_size
+        supervision_segments = torch.stack(
+            (
+                torch.tensor(range(input_lengths.shape[0])),
+                torch.zeros(input_lengths.shape[0]),
+                input_lengths,
+            ),
+            1,
+        ).to(torch.int32)
+
+        dense_fsa_vec = k2.DenseFsaVec(log_probs, supervision_segments)
+
+        decoding_graph = self.graph_compiler.compile(targets, target_lengths)
+        if self.device != log_probs.device:
+            decoding_graph = decoding_graph.to(log_probs.device)
+
+        target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 100.0)
+        tot_scores = target_graph.get_tot_scores(
+            log_semiring=True, use_double_scores=False
+        )
+        if self.reduction == "none":
```

Comment:
What are the differences between `none` and `sum`?
Did you forget to use an early `return` inside the `if` statement?

Reply:
These two different types are just different loss normalization method. There is no logic problem.